backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/CassandraClusterExtension.java

```diff
@@ -43,10 +46,22 @@ public void beforeAll(ExtensionContext extensionContext) {
         Class<?> testClass = extensionContext.getRequiredTestClass();
         if (testClass.getEnclosingClass() == null) {
             cassandraExtension.beforeAll(extensionContext);
-            cassandraCluster = CassandraCluster.create(cassandraModule, cassandraExtension.getDockerCassandra().getHost());
+            start();
         }
     }
 
+    private void start() {
+        cassandraCluster = CassandraCluster.create(cassandraModule, cassandraExtension.getDockerCassandra().getHost());
+    }
+
+    @Override
+    public void beforeEach(ExtensionContext extensionContext) throws Exception {
+        DockerCassandraSingleton.incrementTestsPlayed();
+        DockerCassandraSingleton.restartAfterMaxTestsPlayed(
+            cassandraCluster::close,
+            Throwing.runnable(this::start));
```

Comment:
I like that but I have a little concern. We don't force a restart in every module for cassandra atm, some classes don't use the restart. Which is why there was an extra extension in the first place, and also the `allowRestart` which was added in the `Rule` as well. With this we force the restart everywhere we use the Cassandra extension though... 

My question is, is this an acceptable change? or should we implement something similar to the Rule with `allowRestart`?

Reply:
> My question is, is this an acceptable change?

As a test writter I should not have to know if, or if not, my cassandra server should be restarted. 

Pushing that knowledge in the extension looks good to me.

> or should we implement something similar to the Rule with allowRestart?

Looks like added complexity to me. 

Your tests should not have to care about the concerns and added complexity of running on the same Cassandra server IMO.